flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/internal/executor/BufferReduceStatementExecutor.java

```diff
@@ -0,0 +1,108 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.jdbc.internal.executor;
+
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.types.RowKind;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Currently, this statement executor is only used for table/sql to buffer insert/update/delete
+ * events, and reduce them in buffer before submit to external database.
+ */
+public class BufferReduceStatementExecutor implements JdbcBatchStatementExecutor<RowData> {
+
+	private final JdbcBatchStatementExecutor<RowData> upsertExecutor;
+	private final JdbcBatchStatementExecutor<RowData> deleteExecutor;
+	private final Function<RowData, RowData> keyExtractor;
+	private final Function<RowData, RowData> valueTransform;
+	// the mapping is [KEY, <+/-, VALUE>]
+	private final Map<RowData, Tuple2<Boolean, RowData>> reduceBuffer = new HashMap<>();
+
+	public BufferReduceStatementExecutor(
+			JdbcBatchStatementExecutor<RowData> upsertExecutor,
+			JdbcBatchStatementExecutor<RowData> deleteExecutor,
+			Function<RowData, RowData> keyExtractor,
+			Function<RowData, RowData> valueTransform) {
+		this.upsertExecutor = upsertExecutor;
+		this.deleteExecutor = deleteExecutor;
+		this.keyExtractor = keyExtractor;
+		this.valueTransform = valueTransform;
+	}
+
+	@Override
+	public void prepareStatements(Connection connection) throws SQLException {
+		upsertExecutor.prepareStatements(connection);
+		deleteExecutor.prepareStatements(connection);
+	}
+
+	@Override
+	public void addToBatch(RowData record) throws SQLException {
+		RowData key = keyExtractor.apply(record);
+		boolean flag = changeFlag(record.getRowKind());
+		RowData value = valueTransform.apply(record); // copy or not
+		reduceBuffer.put(key, Tuple2.of(flag, value));
+	}
+
+	/**
+	 * Returns true if the row kind is INSERT or UPDATE_AFTER,
+	 * returns false if the row kind is DELETE or UPDATE_BEFORE.
+	 */
+	private boolean changeFlag(RowKind rowKind) {
+		switch (rowKind) {
+			case INSERT:
+			case UPDATE_AFTER:
+				return true;
+			case DELETE:
+			case UPDATE_BEFORE:
+				return false;
+			default:
+				throw new UnsupportedOperationException(
+					String.format("Unknown row kind, the supported row kinds is: INSERT, UPDATE_BEFORE, UPDATE_AFTER," +
+						" DELETE, but get: %s.", rowKind));
+		}
+	}
+
+	@Override
+	public void executeBatch() throws SQLException {
+		// TODO: reuse the extracted key and avoid value copy in the future
+		for (Tuple2<Boolean, RowData> tuple : reduceBuffer.values()) {
+			if (tuple.f0) {
+				upsertExecutor.addToBatch(tuple.f1);
+			} else {
+				deleteExecutor.addToBatch(tuple.f1);
+			}
+		}
+		upsertExecutor.executeBatch();
+		deleteExecutor.executeBatch();
```

Comment:
Hi @wuchong , may I know the reason why "deleteExecutor.executeBatch()" is placed after "upsertExecutor.executeBatch()"? I have found that the deleteExecutor would accidentally delete previously updated records, resulting in loss of data.

Reply:
There is no matter which one is execute first, because there is no duplicate elements between `upsertExecutor` and `deleteExecutor`.

Reply:
Thanks @wuchong for the timely comments. However, I have observed that some records in `deleteExecutor` contains the same primary key as the records in `upsertExecutor`, and so they are eventually deleted from the database and never appear thereafter.  This problem only happens when "connector.write.flush.max-rows" is larger than 1, so I wonder if this could be a bug related to the bulk flush logic?

Thanks again for your attention : )

Reply:
And by the way, by exchanging the order of the two lines of code mentioned above, the data loss problem also disappears, which looks really strange.

Reply:
> However, I have observed that some records in deleteExecutor contains the same primary key as the records in upsertExecutor.

Are you sure about this? Because the delete records and upsert records are deduplicated by keys in a `Map<KEY, Tuple>` structure. It's hard to believe there are duplicate records with same key. 

Btw, Did you try Flink v1.12? We have fixed a problem with upsert statement in FLINK-15728.

Reply:
Hi @wuchong , glad to see the reply : ) 

After careful verification, I have found you are right that the two lines are actually not the culprit. Rather, the problem disappeared beause I previously filtered (removed) all of the DELETE messages from the connector, as they are always misplaced at the end of all records during bulk operation, thus causing data loss.

Which is to say, when `connector.write.flush.max-rows` is set to 0 or 1, the DELETE messages are correctly put ahead of corresponding UPDATE messages. But as the bulk increases, DELETE messages are executed AFTER all UPDATE messages, and it seems that this reorder issue is not related to [FLINK-15728](https://issues.apache.org/jira/browse/FLINK-15728) , so may I know if there are other reasons that might cause this abnormality?

Thanks : )

Reply:
When the bulk increase, DELETE messages and UPDATE messages will be merged into the `reducedBuffer` first. So there is no matter which one is executed first

Reply:
Yes, and eventually I realized that legacy connector options introduced in Flink 1.10  (i.e. 'connector.type' ) could not enable this feature, instead Flink would use JdbcBatchingOutputFormat which is the reason why the data is incorrect. After replacing the WITH parameters with current ones, the problem solved.

Thanks again for your great effort : )

Reply:
Yes. 'connector.type' is the legacy connector and shouldn't be used anymore.